### PR TITLE
Extension/controllers

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

File: db/migrate/20230418112448_add_role_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUsers < ActiveRecord::Migration[7.0]
+  def change
+    add_column :users, :role, :string, default: 'user'
```
Comment (himanshumishra31) [inline db/migrate/20230418112448_add_role_to_users.rb]: implement it via enum

File: lib/tasks/make_admin.rake [hunk 2]
```diff
@@ -0,0 +1,10 @@
+task :make_admin, [:email] do |t, args|
+  require_relative "../../config/environment.rb"
+  user = User.find_by(email: args[:email])
+  if user.present?
+    user.update(role: "admin")
```
Comment (himanshumishra31) [inline lib/tasks/make_admin.rake]: change it along with enum. Use symbol to update enum
`user.update(role: :admin)`

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -8,7 +8,8 @@ def create
     user = User.find_by(name: params[:name])
     if user.try(:authenticate, params[:password])
       session[:user_id] = user.id
-      redirect_to admin_url
+      user.update_column(:last_activity_time, Time.current)
+      redirect_to user.role == "admin" ? admin_reports_path : admin_url
```
Comment (himanshumishra31) [inline app/controllers/sessions_controller.rb]: `user.admin?`use enum methods after using enum

File: app/controllers/admin/reports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Admin::ReportsController < AdminController
+  def index
+    if params[:from].present? && params[:to].present?
+      start_date = Date.parse(params[:from])
+      end_date = Date.parse(params[:to])
+      @orders = Order.where(created_at: start_date.beginning_of_day..end_date.end_of_day)
+    else
```
Comment (himanshumishra31) [inline app/controllers/admin/reports_controller.rb]: handle only of them is present?

File: app/controllers/admin_controller.rb [hunk 5]
```diff
@@ -1,5 +1,18 @@
...
+  before_action :authenticate_admin!
+
   def index
     @total_orders = Order.count
   end
+
+  private
+
+  def authenticate_admin!
+    redirect_to store_index_path, notice: "You don't have privilege to access this section" unless @user.role == "admin"
```
Comment (himanshumishra31) [inline app/controllers/admin_controller.rb]: translation
enum methods

File: app/controllers/admin_controller.rb [hunk 6]
```diff
@@ -1,5 +1,18 @@
...
+  end
+
+  def set_user
+    @user = User.find(session[:user_id])
```
Comment (himanshumishra31) [inline app/controllers/admin_controller.rb]: use find_by instead of find

Comment (himanshumishra31) [reply app/controllers/admin_controller.rb]: Study difference between find and find_by

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -21,4 +31,29 @@ def set_i18n_locale_from_params
...
+    session[:hit_counter] ||= 0
+    session[:hit_counter] += 1
+    @hit_counter = session[:hit_counter]
+  end
+
+  def set_start_time
+    @start_time = Time.now
+  end
+
+  def add_response_time
```
Comment (himanshumishra31) [inline app/controllers/application_controller.rb]: use around action

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -21,4 +31,29 @@ def set_i18n_locale_from_params
...
+    time_taken = (Time.now - @start_time) * 1000
+    response.headers["x-responded-in"] = "#{time_taken}ms"
+  end
+
+  def check_user_activity_time
+    return unless current_user
+    if current_user.last_activity_time < 10.seconds.ago
```
Comment (himanshumishra31) [inline app/controllers/application_controller.rb]: 5 mins

**Review by himanshumishra31 — CHANGES_REQUESTED** [review]

File: app/controllers/admin_controller.rb (hunk 5 above)
Comment (rishabh-vinsol) [reply app/controllers/admin_controller.rb]: Will do the translation in the next extension 'Mailers and I18n'

File: db/migrate/20230418112448_add_role_to_users.rb (hunk 1 above)
Comment (rishabh-vinsol) [reply db/migrate/20230418112448_add_role_to_users.rb]: Done

File: lib/tasks/make_admin.rake (hunk 2 above)
Comment (rishabh-vinsol) [reply lib/tasks/make_admin.rake]: Done

File: app/controllers/sessions_controller.rb (hunk 3 above)
Comment (rishabh-vinsol) [reply app/controllers/sessions_controller.rb]: Done

File: app/controllers/admin/reports_controller.rb (hunk 4 above)
Comment (rishabh-vinsol) [reply app/controllers/admin/reports_controller.rb]: Done

File: app/controllers/admin_controller.rb (hunk 6 above)
Comment (rishabh-vinsol) [reply app/controllers/admin_controller.rb]: Done

**Pull request merged**